src/main/java/net/minecraftforge/logging/CrashReportAnalyser.java

```diff
@@ -0,0 +1,233 @@
+    }
+
+
+    /**
+     * Iterates over all loaded mods, resolving and caching their package names with the corresponding {@link IModInfo}.
+     * **/
+    public static void cacheModList() {
+        ModList modList = ModList.get();
+        ModuleLayer gameLayer = FMLLoader.getGameLayer();
+
+        if (modList != null) {
+            modList.getMods().forEach(iModInfo -> {
+                //Don't cache minecraft or forge as they will always be included in the stacktrace
+                if (!iModInfo.getModId().equals("forge") && !iModInfo.getModId().equals("minecraft")) {
+                    Set<String> packages = new HashSet<>();
+
+                    gameLayer.findModule(iModInfo.getModId()).ifPresent(module -> packages.addAll(module.getPackages()));
+
+                    packages.forEach(s ->  PACKAGE_MOD_CACHE.put(s, iModInfo));
+                }
+            });
+        }
+    }
+
+    private static void identifyByClass(StackTraceElement stackTraceElement) {
+        blameIfPresent(stackTraceElement);
+    }
+
+    private static void identifyByMixins(StackTraceElement stackTraceElement) {
+        IMixinInfo mixinInfo = getMixinInfo(stackTraceElement);
+
+        if (mixinInfo != null) {
+            String elementAsString = stackTraceElement.toString();
+            String mixinClassName = mixinInfo.getClassName();
+            List<String> targetClasses = mixinInfo.getTargetClasses();
+            blameIfPresent(mixinClassName, "Mixin class: ", mixinClassName, "\n\t\tTarget", (targetClasses.size() == 1 ? ": " + targetClasses.get(0) : "s: " + targetClasses).replaceAll("/", "."), "\n\t\tat ", elementAsString);
+        }
+    }
+
+    private static void blameIfPresent(StackTraceElement stackTraceElement) {
+        blameIfPresent(stackTraceElement.getClassName(), "at ", stackTraceElement.toString());
+    }
+    
+    private static void blameIfPresent(String className, String... position) {
+        String commonPackage = findMatchingPackage(className);
+
+        if (commonPackage != null) {
+            SUSPECTED_MODS.putIfAbsent(PACKAGE_MOD_CACHE.get(commonPackage), position);
+        }
+    }
+
+    /**
+     * Checks if the beginning of the given class name is contained in the {@link #PACKAGE_MOD_CACHE}
+     * @return the  package name of the class or null if it is not present in the cache
+     * */
+    @Nullable
+    private static String findMatchingPackage(String className) {
+        for (String s : PACKAGE_MOD_CACHE.keySet()) {
+            if (className.startsWith(s)) {
+                return s;
+            }
+        }
+
+        return null;
+    }
+
+    @Nullable
+    private static MixinMerged findMixinMerged(StackTraceElement element) {
+        try {
+            Class<?> clazz = Class.forName(element.getClassName());
+            Method[] methods = clazz.getDeclaredMethods();
+            for (Method method : methods) {
+                if (method.getName().equals(element.getMethodName())) {
```

Comment:
Out of interest, what's the difference between this and using `clazz.getDeclaredMethod(element.getMethodName())`?

Reply:
The difference is that we don't know the params of the method and `clazz.getDeclaredMethod()` requires to pass in the classes of the parameters of the method, hence we can only check for the method's name.